### PR TITLE
refactor(core): extract password-validator

### PR DESCRIPTION
**Pull request by simeng-li**
## Summary

Extract a new `PasswordValidator` that will handle all the new password validation and encryption logic.
Make other validators like `SignInExperienceValidator` and `ProfileValidator` more pure. 

No extra logic added. 
 

## Testing

integration test covered

## Checklist

- [ ] `.changeset`
- [ ] unit tests
- [ ] integration tests
- [x] necessary TSDoc comments

### CODE REVIEW

File: packages/core/src/routes/experience/classes/validators/password-validator.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      password,
+      this.passwordPolicyChecker.policy.rejects.userInfo ? userInfo : {}
+    );
+
+    if (issues.length > 0) {
+      throw new RequestError({ code: 'password.rejected', status: 422 }, { issues });
+    }
+  }
+
+  public async encryptPassword(password: string) {
```
Comment (wangsijie) [inline packages/core/src/routes/experience/classes/validators/password-validator.ts:54]: nit: this is an existing issue, we don't actually "encrypt" password, rename to `createPasswordDigest` would be more accurate.

**Review by wangsijie — APPROVED** [review]

File: packages/core/src/routes/experience/classes/validators/password-validator.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+  constructor(
+    private readonly passwordPolicy: SignInExperience['passwordPolicy'],
+    private readonly user?: User
+  ) {
+    this.passwordPolicyChecker = new PasswordPolicyChecker(passwordPolicy);
+  }
+
+  /**
+   * Validate password against the given password policy
```
Comment (gao-sun) [inline packages/core/src/routes/experience/classes/validators/password-validator.ts:35]: the given password policy?

File: packages/core/src/routes/experience/classes/validators/password-validator.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+   * throw a {@link RequestError} 422 if the password is invalid; otherwise, do nothing.
```
Comment (gao-sun) [inline packages/core/src/routes/experience/classes/validators/password-validator.ts:36]:
```suggestion
   * @throws {RequestError} with status 422 if the password is invalid.
```

File: packages/core/src/routes/experience/classes/validators/password-validator.ts (hunk 1 above)
Comment (simeng-li) [reply packages/core/src/routes/experience/classes/validators/password-validator.ts]: renamed

File: packages/core/src/routes/experience/classes/validators/password-validator.ts (hunk 2 above)
Comment (simeng-li) [reply packages/core/src/routes/experience/classes/validators/password-validator.ts]: Password policy is set in SIE.  Anyway, remove the words.

File: packages/core/src/routes/experience/classes/validators/password-validator.ts (hunk 3 above)
Comment (simeng-li) [reply packages/core/src/routes/experience/classes/validators/password-validator.ts]: updated

**Pull request merged**